src/data_source/storage/sql.rs

```diff
@@ -408,6 +428,157 @@ impl SqlStorage {
     }
 }
 
+impl PrunerConfig for SqlStorage {
+    fn set_pruning_config(&mut self, cfg: PrunerCfg) {
+        self.pruner_cfg = Some(cfg);
+    }
+
+    fn get_pruning_config(&self) -> Option<PrunerCfg> {
+        self.pruner_cfg.clone()
+    }
+}
+
+impl SqlStorage {
+    async fn get_minimum_height(&self) -> QueryResult<Option<u64>> {
+        let row = self
+            .query_one_static("SELECT MIN(height) as height FROM header")
+            .await?;
+
+        let height = row.get::<_, Option<i64>>(0).map(|h| h as u64);
+
+        Ok(height)
+    }
+
+    async fn get_height_by_timestamp(&self, timestamp: i64) -> QueryResult<Option<u64>> {
+        let row = self
+            .query_opt(
+                "SELECT height FROM header WHERE timestamp <= $1 ORDER BY height DESC LIMIT 1",
+                [&timestamp],
+            )
+            .await?;
+
+        let height = row.map(|row| row.get::<_, i64>(0) as u64);
+
+        Ok(height)
+    }
+}
+
+#[async_trait]
+impl PrunedHeightStorage for SqlStorage {
+    type Error = QueryError;
+    async fn save_pruned_height(&mut self, height: u64) -> Result<(), Self::Error> {
+        // id is set to 1 so that there is only one row in the table.
+        // height is updated if the row already exists.
+        self.query_opt(
```

Comment:
Can you use `self.upsert` here? It has a bunch of retry logic and error handling stuff baked in